File: blind75/04_product_of_array_except_self/product_of_array_except_self.go
package product_of_array_except_self

// ProductExceptSelf returns an array where each element is the product of all elements except itself.
// Uses prefix-suffix product approach with O(n) time and O(1) extra space.
func ProductExceptSelf(nums []int) []int {
	n := len(nums)
	result := make([]int, n)

	// First pass: fill result with prefix products
	prefix := 1
	for i := 0; i < n; i++ {
		result[i] = prefix
		prefix *= nums[i]
	}

	// Second pass: multiply by suffix products
	suffix := 1
	for i := n - 1; i >= 0; i-- {
		result[i] *= suffix
		suffix *= nums[i]
	}

	return result
}

File: blind75/04_product_of_array_except_self/product_of_array_except_self.py
from typing import List


def product_except_self(nums: List[int]) -> List[int]:
    """
    Returns an array where each element is the product of all elements except itself.
    Uses prefix-suffix product approach with O(n) time and O(1) extra space.

    Args:
        nums: List of integers

    Returns:
        List where each element is the product of all other elements
    """
    n = len(nums)
    result = [1] * n

    # First pass: fill result with prefix products
    prefix = 1
    for i in range(n):
        result[i] = prefix
        prefix *= nums[i]

    # Second pass: multiply by suffix products
    suffix = 1
    for i in range(n - 1, -1, -1):
        result[i] *= suffix
        suffix *= nums[i]

    return result
